### PR TITLE
Alternative approach to typed ArgumentInformation

### PR DESCRIPTION
**Pull request by Shane32**
This is the only alternative I would consider.  It reduces the amount of code by a mere 30 lines.  I don't see how this is a benefit to the project.

### CODE REVIEW

File: src/GraphQL/FromServicesAttribute.cs [hunk 1]
```diff
@@ -10,11 +10,8 @@ namespace GraphQL
...
         /// <inheritdoc/>
-        public override void Modify<TParameterType>(ArgumentInformation<TParameterType> argumentInformation)
-            => argumentInformation.Expression = context => GetRequiredService<TParameterType>(context);
-
-        private static TServiceType GetRequiredService<TServiceType>(IResolveFieldContext context)
-            => (context.RequestServices ?? throw new MissingRequestServicesException())
-                .GetRequiredService<TServiceType>();
+        public override void Modify<TParameterType>(ArgumentInformation argumentInformation)
+            => argumentInformation.SetExpression(context => (context.RequestServices ?? throw new MissingRequestServicesException())
+                .GetRequiredService<TParameterType>());
```
Comment (Shane32) [inline src/GraphQL/FromServicesAttribute.cs:15]: In order to be able to EASILY set a type-safe expression, without having to write `(Expression<Func<IResolveFieldContext, TParameterType>>)`, I added a generic method called `SetExpression`.  I also made this method accept a delegate, which is honestly more flexible in its use, as shown here where we can use null coalescing operators, code blocks, and so on.

File: src/GraphQL/GraphQLAttribute.cs [hunk 2]
```diff
@@ -57,7 +57,7 @@ public virtual void Modify(TypeInformation typeInformation)
         /// <summary>
         /// Updates the properties of the specified <see cref="ArgumentInformation{TParameterType}"/> as necessary.
         /// </summary>
-        public virtual void Modify<TParameterType>(ArgumentInformation<TParameterType> argumentInformation)
+        public virtual void Modify<TParameterType>(ArgumentInformation argumentInformation)
```
[the remark this answers is not in the document]
Comment (sungam3r) [reply src/GraphQL/GraphQLAttribute.cs:60]: I understand your intentions. Even with generic argument one can write invalid expression using invalid cast or something else.

File: src/GraphQL/FromServicesAttribute.cs (hunk 1 above)
Comment (sungam3r) [reply src/GraphQL/FromServicesAttribute.cs:15]: > to be able to EASILY set a type-safe expression

OK. I'm also want to end with the most easy solution. Consider `HardcodedValueAttribute` from my latest commit. One wants to return hardcoded value 85. So he should to cast 85 to `TParameterType` , right? Something like `context => (TParameterType)(object)85`. Do you mean that this syntax is simple? Type-safety is ephemeral here. In my variant the code looks like `context => 85`.

Comment (Shane32) [reply src/GraphQL/FromServicesAttribute.cs:15]: Exactly -- you are hiding the inherit loss of type safety.  It should be explicit.

**Comment by Shane32** [comment]
There is a third approach: we could move the call to the typed `Modify` inside the attribute.  Perhaps with a derived attribute class.

**Comment by Shane32** [comment]
> There is a third approach: we could move the call to the typed `Modify` inside the attribute. Perhaps with a derived attribute class.

If you want I can add another PR with this approach demonstrated.

**Comment by sungam3r** [comment]
If I understand correctly, then all 3 of the proposed options work. The difference is only as we understand _type safety_. More precisely, even as we **evaluate** _type safety_. I understand that you want to provide easy means to set expression in _type safe manner_. OK, good goal. I'm trying to convey the idea that this simplicity is actually ephemeral and in practice does not carry a tangible benefit, only complicating our own code jumping over generics. Also in some cases that _type safety_ complicates user code as in `HardcodedValueAttribute` example.

**Comment by sungam3r** [comment]
> we could move the call to the typed Modify inside the attribute. 

Hmmm, `Modify` is already a method inseide attrubute, I don't get it.

**Comment by Shane32** [comment]
If you say so, then that's what I want, yes.

> Also in some cases that type safety complicates user code as in HardcodedValueAttribute example.

Then we can look for another approach to that example.  Or let it be.

**Comment by sungam3r** [comment]
By the way, is it already so critical that my variant of code is not so _type safe_? After all error will occur at schema initialization stage, i.e. on app startup. It is practically the same as when compiling, in a certain sense. I think, you understood what I mean. We have a lot of such checks.

File: src/GraphQL/FromServicesAttribute.cs (hunk 1 above)
Comment (sungam3r) [reply src/GraphQL/FromServicesAttribute.cs:15]: 😕  Then there should be unwrapping code for `Convert` expression in your variant for such cases if you really want to eliminate boxing and achive the same performance as with direct method call.

**Comment by sungam3r** [comment]
> If you want I can add another PR with this approach demonstrated.

😄 I want to achive consensus.

**Comment by Shane32** [comment]
- #2916

**Comment by Shane32** [comment]
> 😄 I want to achive consensus.

I'm trying the best I can.  I do not wish to lose generics support or type safety because of "simplification".  I'm trying to offer alternatives.

**Comment by Shane32** [comment]
Replaced by #2916

**Pull request closed without merging**